src/discord/dto/read-discord.dto.ts

```diff
@@ -0,0 +1,6 @@
+import { PartialType } from '@nestjs/mapped-types';
+import { CreateDiscordDto } from './create-discord.dto';
+
+export class ReadDiscordDto extends PartialType(CreateDiscordDto) {
+  id: number;
```

Comment:
is `id` needed if it is created from `CreateDiscordDto`?

Reply:
This line no `5` is making it necessary, `id` exist with that line too but it is optional ( `id?: number` ).

Reply:
I do not understand, where is it optional? It looks like it is required here...? https://github.com/EddieJaoudeCommunity/api/pull/23/files#diff-cdd95909b05df4daa83d67707b55752498545cda37904ddf279494e56aa4bedeR2

Reply:
I mean to say when we use `PartialType`, it will copy the values and make all of them optional (excluding nested).

Here, we have these properties in `create-discord.dto.ts`
```ts
  id: number;
  username: string;
  bio: {
    description?: string;
    twitter?: string;
    linkedin?: string;
    github?: string;
  };
  updatedOn: Date;
  createdOn: Date;
```

The top-level properties are necessary but in `read-discord.dto.ts`
If I just add
```ts
export class ReadDiscordDto extends PartialType(CreateDiscordDto) {}
```

This will results in `ReadDiscordDto` having the following properties
```ts
  id?: number;
  username?: string;
  bio?: {
    description?: string;
    twitter?: string;
    linkedin?: string;
    github?: string;
  };
  updatedOn?: Date;
  createdOn?: Date;
```

But since `id` is a primary key, which is required in every **API**, so we need to make it necessary. To do that, we can redefine that specific property in the `ReadDiscordDto`
```ts
export class ReadDiscordDto extends PartialType(CreateDiscordDto) {
  id: number;
}
```

Now this will result in
```ts
  id: number;
  username?: string;
  bio?: {
    description?: string;
    twitter?: string;
    linkedin?: string;
    github?: string;
  };
  updatedOn?: Date;
  createdOn?: Date;
```

---
In short, whatever property we define in that class will be compulsory. `PartialType` copy the properties and make it optional, if we need to make any property compulsory, we have to redefine it.

Reply:
Ah I see, thank you for clarifying 👍